# Predicting Credit Card Payment Default

## Notebook 1: Cleaning Data and Pre-Processing
__Steps to load, clean, and analyze the data upon first inspection.__
***

## Table of Contents
1. [**Introduction**](#a1)<br>
    1.1 [Business Question](#a1.1)<br>
    1.2 [Data Collection](#a1.2)<br>
    
2. [**Loading & Checking**](#a2)<br>
    2.1 [Data and Column Descriptions](#a2.1)<br>
    2.2 [Data Dictionary](#a2.2)

3. [**Investigate Columns**](#a3)<br>
    3.1.[Correcting Column names](#a3.1)<br>
    3.2.[Clean unknown values ](#a3.2)<br>

# Introduction <a id="a1"></a>


### Business Question <a id="a1.1"></a>

#### Can we accurately predict the probability of a customer defaulting on their payments?

### Context 

Credit card are issued by banks or financial companies which allow clients to borrow money to pay for goods or services. Credit card companies maintain vast databases on cardholders and have the condition that cardholders pay back the company on time. Since the companies make the payment on the clients behalf it is in thier best interest that they get thier money back from the cardholders. 

There are a number of ways the issuers makes thier profit but it mainly comes from annual fees to cardholders, transaction fees on purchases or transfers, processor fees but the largest profit portion comes from interest fees. Interest fees are charged when balance on the card is not paid on time. To read more on how these fees are calculated you can read more on the link. [How do credit card companies make money](https://www.fool.com/the-ascent/credit-cards/how-do-credit-card-companies-make-money/)

The bottom line is that the fees are calculated based on credit risk which is determined by credit history of individual. However with the case of this particular dataset the reality is quite different.

### Data Background 

In order to encourage the flow of capital, banks in Taiwan began extensively promoting the population to apply for credit cards. They lowered the requirements for approvals mainly targeting young people and students who lacked practical experience and financial knowledge, had no job or lower income. 
By February 2006 Taiwan had a credit card debt of $268 billion USD! This led to a range of social issues such as homelessness, violent debt collection especially towards lower income families and even suicide just because they were not able to pay off the debt. Until the government stepped in to sort the problems and implemented stricter rules that monitored who could be approved for credit cards as well as how much limit they can get.
This is how we get our main problem statement and business motivation:

How can we accurately predict the probability of a customer defaulting especially when we have very limited to no credit history information?
So, in this classification problem our stakeholders are credit card companies whose main motivation is not the accuracy of the prediction model but rather the recall rate. The recall measures the proportion of actual positives that were identified correctly. We are more interested in the number of false negatives being as low as possible and want the number of true positives to be high as possible. Basically, a customer who has been incorrectly predicted as likely to not default when in reality they are likely to default, poses the biggest risk to the bank.


### Data collection
Normally credit card data is hard to find as it contains private personal details such as marital status, employment status etc. This particular dataset was donated to UCI Machine learning repository, however I initially found it on Kaggle.com. They are both the same but the original on UCI website has a fuller data dictionary of the two.

You can view the original data from [here](https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients)

### Dataset Information
This dataset contains information on default payments, demographic factors, credit data, history of payment, and bill statements of credit card clients in Taiwan from April 2005 to September 2005.

***
__Importing Libraries__

First we must import libraries we need to clean and analyse the data.

In [2]:
#importing libraries we may need
import pandas as pd
import numpy as np

***
## Loading & Checking Data <a id="a2"></a>
### Data and Column Description

Let's take a closer look at the data, before doing any analysis.

In [9]:
#import the dataset from folder
credit_df=pd.read_csv(r'C:\Users\Robin\Downloads\capstone_project\_data\taiwan_data.csv')

In [10]:
#lets look at the shape of the dataframe
print(f'We have {credit_df.shape[0]} rows and {credit_df.shape[1]} columns in the dataset.')

We have 30000 rows and 25 columns in the dataset.


In [3]:
#a snapshot of the first few rows
credit_df.head() 

NameError: name 'credit_df' is not defined

In [12]:
credit_df.tail() # last 5 rows

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
29995  29996     220000    1          3         1   39      0      0      0   
29996  29997     150000    1          3         2   43     -1     -1     -1   
29997  29998      30000    1          2         2   37      4      3      2   
29998  29999      80000    1          3         1   41      1     -1      0   
29999  30000      50000    1          2         1   46      0      0      0   

       PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
29995      0  ...      88004      31237      15980      8500     20000   
29996     -1  ...       8979       5190          0      1837      3526   
29997     -1  ...      20878      20582      19357         0         0   
29998      0  ...      52774      11855      48944     85900      3409   
29999      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
29995      5003      3047      5000      1000                           0  
29996      8998       129         0         0                           0  
29997     22000      4200      2000      3100                           1  
29998      1178      1926     52964      1804                           1  
29999      1430      1000      1000      1000                           1  

[5 rows x 25 columns]

*** 

## Data dictionary <a id="a1"></a>
<table>
  <tr>
    <th style="text-align: left; background: lightgrey">Column Name</th>
    <th style="text-align: left; background: lightgrey">Column Contents</th>
  </tr>
  <tr>
    <td style="text-align: left">ID</th>
    <td style="text-align: left">ID of each client</th>
  <tr>
  <tr>
    <td style="text-align: left">LIMIT_BAL</th>
    <td style="text-align: left">Amount of given credit in NT dollars (includes individual and family/supplementary credit)</th>
  <tr>
<tr>
    <td style="text-align: left">SEX</th>
    <td style="text-align: left">Gender (1=male, 2=female)</th>
  <tr>
<tr>
    <td style="text-align: left">EDUCATION</th>
    <td style="text-align: left">Level of education(1=graduate school, 2=university, 3=high school, 4=others, 5=unknown, 6=unknown)</th>
  <tr>
<tr>
    <td style="text-align: left">MARRIAGE</th>
    <td style="text-align: left">Marital status (1=married, 2=single, 3=others)</th>
  <tr>
<tr>
    <td style="text-align: left">AGE</th>
    <td style="text-align: left">Age in years</th>
  <tr>
  <tr>
    <td style="text-align: left">PAY_0</th>
    <td style="text-align: left">Repayment status in September, 2005 (-1=pay duly, 1=payment delay for one month, 2=payment delay for two months, 
    … 8=payment delay for eight months, 9=payment delay for nine months and above)</th>
  <tr>
  <tr>
    <td style="text-align: left">PAY_2</th>
    <td style="text-align: left">Repayment status in August, 2005 (scale same as above)</th>
  <tr>
  <tr>
    <td style="text-align: left">PAY_3</th>
    <td style="text-align: left">Repayment status in July, 2005 (scale same as above)</th>
  <tr>
<tr>
    <td style="text-align: left">PAY_4</th>
    <td style="text-align: left">Repayment status in June, 2005 (scale same as above)</th>
  <tr>
  <tr>
    <td style="text-align: left">PAY_5</th>
    <td style="text-align: left">Repayment status in May, 2005 (scale same as above)</th>
  <tr>
  <tr>
    <td style="text-align: left">PAY_6</th>
    <td style="text-align: left">Repayment status in April, 2005 (scale same as above)</th>
  <tr>
<tr>
    <td style="text-align: left">BILL_AMT1</th>
    <td style="text-align: left">Amount of bill statement in September, 2005 (NT dollar)</th>
  <tr>
<tr>
    <td style="text-align: left">BILL_AMT2</th>
    <td style="text-align: left">Amount of bill statement in August, 2005 (NT dollar)</th>
  <tr>
  <tr>
    <td style="text-align: left">BILL_AMT3</th>
    <td style="text-align: left">Amount of bill statement in July, 2005 (NT dollar)</th>
  <tr>
  <tr>
    <td style="text-align: left">BILL_AMT4</th>
    <td style="text-align: left">Amount of bill statement in June, 2005 (NT dollar)</th>
  <tr>
  <tr>
    <td style="text-align: left">BILL_AMT5</th>
    <td style="text-align: left">Amount of bill statement in May, 2005 (NT dollar)</th>
  <tr>
  <tr>
    <td style="text-align: left">BILL_AMT6</th>
    <td style="text-align: left">Amount of bill statement in April, 2005 (NT dollar)</th>
  <tr>
  <tr>
    <td style="text-align: left">PAY_AMT1</th>
    <td style="text-align: left">Amount of previous payment in September, 2005 (NT dollar)</th>
  <tr>
 <tr>
    <td style="text-align: left">PAY_AMT2</th>
    <td style="text-align: left">Amount of previous payment in August, 2005 (NT dollar)</th>
  <tr>
   <tr>
    <td style="text-align: left">PAY_AMT3</th>
    <td style="text-align: left">Amount of previous payment in July, 2005 (NT dollar)</th>
  <tr>
 <tr>
    <td style="text-align: left">PAY_AMT4</th>
    <td style="text-align: left">Amount of previous payment in June, 2005 (NT dollar)</th>
  <tr>
   <tr>
    <td style="text-align: left">PAY_AMT5</th>
    <td style="text-align: left">Amount of previous payment in May, 2005 (NT dollar)</th>
  <tr>
 <tr>
    <td style="text-align: left">PAY_AMT6</th>
    <td style="text-align: left">Amount of previous payment in April, 2005 (NT dollar)</th>
  <tr>
 <tr>
    <td style="text-align: left">default.payment.next.month</th>
    <td style="text-align: left">Default payment (1=yes, 0=no)</th>
  <tr>
</table>
<br>

***
## Data cleaning and preprocessing

__Dataset Summary Information__

In [14]:
#check for nulls,missing values and duplicates

credit_df.isna().sum()

ID                            0
LIMIT_BAL                     0
SEX                           0
EDUCATION                     0
MARRIAGE                      0
AGE                           0
PAY_0                         0
PAY_2                         0
PAY_3                         0
PAY_4                         0
PAY_5                         0
PAY_6                         0
BILL_AMT1                     0
BILL_AMT2                     0
BILL_AMT3                     0
BILL_AMT4                     0
BILL_AMT5                     0
BILL_AMT6                     0
PAY_AMT1                      0
PAY_AMT2                      0
PAY_AMT3                      0
PAY_AMT4                      0
PAY_AMT5                      0
PAY_AMT6                      0
default payment next month    0
dtype: int64

In [15]:
#check for duplicates
credit_df.duplicated().sum()

0

In [16]:
credit_df.columns
#pd.DataFrame(credit_df.columns)
#looking at columns names to check all the column names are appropriate

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

The columns PAY_0-PAY_6 are not very intuitive columns names. These columns represent the repayment history of the customer, so we can rename them to something like HIST_PAY1.
The sixth column starts at PAY_0 but then skips to PAY_2. Looking at the other columns in a group such as BILL_AMT and PAY_AMT they all start at 1 and go up to 6. So we need to rename the column PAY_0 as PAY_1 so it has the same format as the other columns and avoid confusion. 

Default payment next month is the target column i.e the one we want to predict. This column name should also be changed for easier access,reference and most importantly so it follows the same uniform naming convention as the other columns - 1 or 2 words in capitals that succinctly describes the data contained in the column.

In [17]:
new_column_names = {'PAY_0': 'PAY_1',
        'default payment next month':'DEFAULT'
        }

In [18]:
credit_df.rename(columns=new_column_names,inplace=True)

In [19]:
# check they have been changed and also the datatypes
credit_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   ID         30000 non-null  int64
 1   LIMIT_BAL  30000 non-null  int64
 2   SEX        30000 non-null  int64
 3   EDUCATION  30000 non-null  int64
 4   MARRIAGE   30000 non-null  int64
 5   AGE        30000 non-null  int64
 6   PAY_1      30000 non-null  int64
 7   PAY_2      30000 non-null  int64
 8   PAY_3      30000 non-null  int64
 9   PAY_4      30000 non-null  int64
 10  PAY_5      30000 non-null  int64
 11  PAY_6      30000 non-null  int64
 12  BILL_AMT1  30000 non-null  int64
 13  BILL_AMT2  30000 non-null  int64
 14  BILL_AMT3  30000 non-null  int64
 15  BILL_AMT4  30000 non-null  int64
 16  BILL_AMT5  30000 non-null  int64
 17  BILL_AMT6  30000 non-null  int64
 18  PAY_AMT1   30000 non-null  int64
 19  PAY_AMT2   30000 non-null  int64
 20  PAY_AMT3   30000 non-null  int64
 21  PAY_AMT4   3

*** 
### Pre-Processing
Lets focus on cleaning and checking for inconsistencies in the demographic columns of the dataset first.

__Column: SEX__ <br>

From the data dictionary we already know that sex has the values 1 and 2 representing male and female. However in computing binary values are traditionally represented as 0 and 1, so we will change them.



In [4]:
credit_df['SEX'].unique()

NameError: name 'credit_df' is not defined

In [21]:
credit_df['SEX']=credit_df['SEX'].replace({1: 0, 2: 1}) 

In [22]:
credit_df['SEX'].value_counts() # check we only have 0 and 1 left

1    18112
0    11888
Name: SEX, dtype: int64

***
__Column: Education__

Education also has some extra values. We already know there is 2 redundent values of 5 and 6. We have 4 to denote 'unknown' or 'other' education status. We can do a value counts to check if there are additional unknown values and if there are then we can also assign them to 4.

In [23]:
credit_df['EDUCATION'].value_counts(normalize=True)*100 


2    46.766667
1    35.283333
3    16.390000
5     0.933333
4     0.410000
6     0.170000
0     0.046667
Name: EDUCATION, dtype: float64

In [24]:
# as suspected another unknown value - 0,not mentioned in description
credit_df['EDUCATION']=credit_df['EDUCATION'].replace([0,5,6], 4 )
credit_df['EDUCATION'].value_counts(normalize=True)*100 
# now do get dummies so we only have 3 columns left

2    46.766667
1    35.283333
3    16.390000
4     1.560000
Name: EDUCATION, dtype: float64

There is actually 3 unknown values: 0,5,6. We can map these to the value 4 to refer to other either instituations or other unknown place of study.

However upon doing some external reseach, I found that Taiwan has a very high percentage of education completion in general. In fact around 94.7% of people have passed high school and gone onto some sort of higher education or senior vocational course . Based on this information, it is safe to assume that the unknown values and other category can be grouped together in the value 3, as it is highly likely that these people have passed high school at very the least.
Additionally since adding the 3 unknown values to 4 gives us 468 values which accounts for less than 1.56% of the entire column it is insignificant in making predictions which means it can only contribute to making models more computationally expensive to compute.

In [25]:
education_df = pd.get_dummies(credit_df['EDUCATION'],prefix='EDUCATION') 

In [26]:
education_df.drop(columns=['EDUCATION_4'], inplace=True)

In [27]:
education_df

Education_1  Education_2  Education_3
0                0            1            0
1                0            1            0
2                0            1            0
3                0            1            0
4                0            1            0
...            ...          ...          ...
29995            0            0            1
29996            0            0            1
29997            0            1            0
29998            0            0            1
29999            0            1            0

[30000 rows x 3 columns]

In [28]:
education_df.value_counts()

Education_1  Education_2  Education_3
0            1            0              14030
1            0            0              10585
0            0            1               4917
                          0                468
dtype: int64

***
__Column: Marriage__

This also has 1 unknown value of 0, which we can change to 3 (others)

In [29]:
credit_df['MARRIAGE'].value_counts()

2    15964
1    13659
3      323
0       54
Name: MARRIAGE, dtype: int64

In [30]:
# random 0 value - not mentioned in description anywhere, lets change it to others

credit_df['MARRIAGE']=credit_df['MARRIAGE'].replace(0,3)

In [31]:
credit_df['MARRIAGE'].value_counts(normalize=True)*100 

2    53.213333
1    45.530000
3     1.256667
Name: MARRIAGE, dtype: float64

Married and single can essentially be represented as 1 columns of married or not. Although value 3 only accounts for 1.25% of the column we can at the moment have a separate column for it but later when we do PCA or feature engineering I suspect it will be unimportant in predictions ans thus will be removed or not included in the modelling.


In [32]:
credit_df['MARRIAGE']

0        1
1        2
2        2
3        1
4        1
        ..
29995    1
29996    2
29997    2
29998    1
29999    1
Name: MARRIAGE, Length: 30000, dtype: int64

In [33]:
marriage_df=pd.get_dummies(credit_df['MARRIAGE'],prefix='Marital_status',drop_first=True)

In [34]:
marriage_df

Marital_status_2  Marital_status_3
0                     0                 0
1                     1                 0
2                     1                 0
3                     0                 0
4                     0                 0
...                 ...               ...
29995                 0                 0
29996                 1                 0
29997                 1                 0
29998                 0                 0
29999                 0                 0

[30000 rows x 2 columns]

We have removed the first column as the second column represents 0 for Married people and 1 for sinlge people. Maritial status_3 represents any another values.
***
__Putting back the cleaned data into 1 DataFrame__

In [36]:
credit_df_clean = pd.concat([credit_df,education_df, marriage_df], axis=1)
credit_df_clean.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  PAY_4  \
0   1      20000    1          2         1   24      2      2     -1     -1   
1   2     120000    1          2         2   26     -1      2      0      0   
2   3      90000    1          2         2   34      0      0      0      0   
3   4      50000    1          2         1   37      0      0      0      0   
4   5      50000    0          2         1   57     -1      0     -1      0   

   ...  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  DEFAULT  Education_1  \
0  ...         0         0         0         0        1            0   
1  ...      1000      1000         0      2000        1            0   
2  ...      1000      1000      1000      5000        0            0   
3  ...      1200      1100      1069      1000        0            0   
4  ...     10000      9000       689       679        0            0   

   Education_2  Education_3  Marital_status_2  Marital_status_3  
0            1            0                 0                 0  
1            1            0                 1                 0  
2            1            0                 1                 0  
3            1            0                 0                 0  
4            1            0                 0                 0  

[5 rows x 30 columns]

In [37]:
#create the new columns names according to what they represent
rename_col = {'Education_1': 'Education_higher',
        'Education_2':'Education_university',
        'Education_3':'Education_highschool',
        'Marital_status_2':'Marriage_Single',
        'Marital_status_3':'Marriage_Other'
        }

In [38]:
credit_df_clean.rename(columns=rename_col,inplace=True)
credit_df_clean.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  PAY_4  \
0   1      20000    1          2         1   24      2      2     -1     -1   
1   2     120000    1          2         2   26     -1      2      0      0   
2   3      90000    1          2         2   34      0      0      0      0   
3   4      50000    1          2         1   37      0      0      0      0   
4   5      50000    0          2         1   57     -1      0     -1      0   

   ...  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  DEFAULT  Education_higher  \
0  ...         0         0         0         0        1                 0   
1  ...      1000      1000         0      2000        1                 0   
2  ...      1000      1000      1000      5000        0                 0   
3  ...      1200      1100      1069      1000        0                 0   
4  ...     10000      9000       689       679        0                 0   

   Education_university  Education_highschool  Marriage_Single  Marriage_Other  
0                     1                     0                0               0  
1                     1                     0                1               0  
2                     1                     0                1               0  
3                     1                     0                0               0  
4                     1                     0                0               0  

[5 rows x 30 columns]

In [39]:
# we can now drop the old columns
credit_df_clean=credit_df_clean.drop(columns=['EDUCATION','MARRIAGE'])

In [40]:
credit_df_clean

ID  LIMIT_BAL  SEX  AGE  PAY_1  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  \
0          1      20000    1   24      2      2     -1     -1     -2     -2   
1          2     120000    1   26     -1      2      0      0      0      2   
2          3      90000    1   34      0      0      0      0      0      0   
3          4      50000    1   37      0      0      0      0      0      0   
4          5      50000    0   57     -1      0     -1      0      0      0   
...      ...        ...  ...  ...    ...    ...    ...    ...    ...    ...   
29995  29996     220000    0   39      0      0      0      0      0      0   
29996  29997     150000    0   43     -1     -1     -1     -1      0      0   
29997  29998      30000    0   37      4      3      2     -1      0      0   
29998  29999      80000    0   41      1     -1      0      0      0     -1   
29999  30000      50000    0   46      0      0      0      0      0      0   

       ...  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  DEFAULT  Education_higher  \
0      ...         0         0         0         0        1                 0   
1      ...      1000      1000         0      2000        1                 0   
2      ...      1000      1000      1000      5000        0                 0   
3      ...      1200      1100      1069      1000        0                 0   
4      ...     10000      9000       689       679        0                 0   
...    ...       ...       ...       ...       ...      ...               ...   
29995  ...      5003      3047      5000      1000        0                 0   
29996  ...      8998       129         0         0        0                 0   
29997  ...     22000      4200      2000      3100        1                 0   
29998  ...      1178      1926     52964      1804        1                 0   
29999  ...      1430      1000      1000      1000        1                 0   

       Education_university  Education_highschool  Marriage_Single  \
0                         1                     0                0   
1                         1                     0                1   
2                         1                     0                1   
3                         1                     0                0   
4                         1                     0                0   
...                     ...                   ...              ...   
29995                     0                     1                0   
29996                     0                     1                1   
29997                     1                     0                1   
29998                     0                     1                0   
29999                     1                     0                0   

       Marriage_Other  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
29995               0  
29996               0  
29997               0  
29998               0  
29999               0  

[30000 rows x 28 columns]

***
__Column: PAY columns values__

The PAY_ columns range is actually -2 to 8 whereas in the dictionary it said -1 to 9. Let us make it 0-10 instead so all values are positive but the range and what it represents is unchanged. 

In [ ]:
credit_df_pay_col=credit_df_clean.columns[4:10]
for month in credit_df_pay_col:
    print(sorted(credit_df_clean[month].unique()))

In [ ]:
# These ranges are not mentioned in the description, let's change the range to back to -1-9?

for month in credit_df_clean.columns[4:10]:
    credit_df_clean[month]=credit_df_clean[month]+2
#                       inplace=True)


credit_df_clean['PAY_1'].value_counts()

In [ ]:
credit_df_clean.head()

*** 
### Saving the work

In [ ]:
#save the cleaned version of the data
credit_df_clean.to_csv('credit_df_dataset_cleaned.csv', index=False) # the index column will not be saved

***
### Ending notes
In this notebook, I have mainly changed column names to more suitable names, reassign unknown values in column. Much of the data was originally categorical data that was transformed into ascending order values. However as this could be misinterpreted as values are ordinal. So re-encoding the categorical values into binary 0 and 1 was essential. 

The next steps are to start doing some inital EDA on the dataset to see what initial insights we can gain.

***

### References

https://sevenpillarsinstitute.org/case-studies/taiwans-credit-card-crisis/
https://www.fool.com/the-ascent/credit-cards/how-do-credit-card-companies-make-money/ 
